Sources/core/objects/Projectiles.cpp

```diff
@@ -0,0 +1,221 @@
+#include "Projectiles.h"
+#include "../Helpers.h"
+
+std::ostream& operator<<(std::stringstream& file,const std::vector<Dir>& s)
+{
+    file << " " << s.size() << " ";
+    for (auto it : s)
+        file << " " << it << " ";
+    return file;
+}
+
+std::istream& operator>>(std::stringstream& file,std::vector<Dir>& s)
+{
+    size_t size;
+    file >> size;
+    s.reserve(size);
+
+    unsigned int local_id;
+    for (size_t i = 0; i < size; ++i)
+    {
+        file >> local_id;
+        s.push_back(local_id);
+    }
+    return file;
+}
+
+unsigned int hash(const std::vector<Dir>& content)
+{
+    unsigned int retval = 0;
+    int i = 1;
+    for (auto it = content.begin(); it != content.end(); ++it, ++i)
+    {
+        retval += *it * i;
+    }
+    return retval;
+}
+
+
+Projectile::Projectile(size_t id) : IMovable(id)
+{
+    SetPassable(D_ALL, Passable::FULL);
+    damage = 0;
+    SetSprite("icons/projectiles.dmi");
+    v_level = 5;
+    current_step_ = 1;
+    tickSpeed = 1;
+    pixSpeed = 4;
+}
+void Projectile::Process()
+{
+    if (movement_.size() == 0)
+    {
+        qDebug() << "Trajectory for the projectile is empty, critical error!";
+        abort();
+    }
+    Dir step =  movement_[current_step_];
+    Rotate(step);
+    if(!CheckPassable())
+    {
+        return;
+    }
+    current_step_ = (current_step_ + 1) % movement_.size();
+    MainMove();
+}
+
+bool Projectile::CheckPassable()
+{
+    auto tile = GetOwner();
+    if (!CanPass(tile->GetPassable(D_ALL), passable_level))
+    {
+        tile->Bump(GetId());
+        Delete();
+        return false;
+    }
+    if (IMovable::CheckPassable())
+    {
+        return true;
+    }
+    Delete();
+    return false;
+}
+//TODO maybe it can apply froce depending on bullet speed but that should be considered        ApplyForce(DirToVDir[m->dMove]);
+void Projectile::MakeMovementPattern(VDir target, Dir facing)
+{
+    int x = target.x;
+    int y = target.y;
+    int abs_x = std::abs(x);
+    int abs_y = std::abs(y);
+    if (abs_x == abs_y)
+    {
+        if (y < 0)
+        {
+            if (facing == 0|| facing == 1)
+            {
```

Comment:
Again, `Dir` variable should be compared only with another `Dir`, so `D_LEFT` and `D_RIGHT`.